---
title: "Gastrulation: local correlation between DNA methylation and RNA expression in EB cells"
output: 
  BiocStyle::html_document: 
    fig_width: 12
    fig_height: 8
---

```{r echo=FALSE, include=FALSE}
suppressPackageStartupMessages(library(scater))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(ggplot2))
```

<!-- Define functions -->
```{r definefuncs, echo=FALSE}
loadMultipleMetData <- function(indir, cells, annotation=NULL) {
  data_list <- list()
  for (cell in cells) {
    if (file.exists(sprintf("%s/%s.tsv.gz",indir,cell))) {
      # print(sprintf("Loading %s...",cell))
      data_list[[cell]] <- fread(sprintf("zcat < %s/%s.tsv.gz",indir,cell), sep="\t", stringsAsFactors=T) %>%
        setnames(c("chr","pos","rate")) %>%
        .[,sample:=cell] %>% .[,c("start","end"):=list(pos,pos)] %>% 
        setkey("chr","start","end") %>%
        foverlaps(annotation, nomatch=0) %>% 
        .[,dist:=ifelse(strand %in% c("+","*"),pos-TSS,TSS-pos)] %>% 
        .[,c("sample","ens_id","dist","rate")]
    }
  }
  return(rbindlist(data_list))
}
```

<!-- Define I/O and options -->
```{r define_opts, echo=FALSE, include=FALSE}

## Define I/O ##
io <- list()
io$basedir   <- "/Users/ricard/data/NMT-seq_EB"
io$outdir <- "/Users/ricard/gastrulation/metrna/local_coupling/EB_cells/out"
io$tss <- "/Users/ricard/data/mm10_regulation/promoters/TSS.bed"
io$in.sample_metadata <- paste0(io$basedir,"/sample_metadata.txt")
io$met.indir <- paste(io$basedir,"met/raw",sep="/")
io$rna.infile <- paste(io$basedir,"rna/parsed/sceset.rds",sep="/")

## Define options ##
opts <- list()

# Define window options
opts$up <- 5000
opts$window <- 300
opts$slide <- 50
opts$down <- 5000

# Define colors for the lineages
opts$colors <- c(Pluripotent="#FFD700", Differentiated="#CD853F")

# Define which cells to use
opts$cells <- fread(io$in.sample_metadata) %>% 
  .[pass_metQC==T & pass_rnaQC==T & culture=="EB",sample] 
```

<!-- Load gene metadata -->
```{r}
# All promoters
tss <- fread(io$tss,stringsAsFactors=TRUE)[,c(1,2,3,6,4)] %>%
  setnames(c("chr","start","end","strand","ens_id")) %>%
  .[,chr:=as.factor(sub("chr","",chr))] %>%
  .[,TSS:=start] %>%
  .[,c("start","end"):=.(start-opts$up,end+opts$down)] %>%
  setkey(chr,start,end)
```

<!-- Load sample metadata and define which cells to use -->
```{r}
sample_metadata <- fread(io$in.sample_metadata, stringsAsFactors=T) %>% 
  .[,c("sample","lineage")] %>%
  .[sample%in%opts$cells]
```

<!-- Load methylation data -->
```{r load_data, echo=FALSE, include=FALSE}
met_dt <- loadMultipleMetData(io$met.indir, opts$cells, tss)
```

<!-- Load RNA expression data -->
```{r}
# Load SCE object
sce <- readRDS(file=io$rna.infile) %>% toSingleCellExperiment()
sce <- sce[,colnames(sce) %in% opts$cells]

# Convert to data.table
rna_dt <- exprs(sce) %>% t %>% as.data.table(keep.rownames="sample") %>% 
  melt(id.vars="sample", value.name="expr", variable.name="gene") %>%
  merge(rowData(sce) %>% as.data.frame(row.names=rownames(sce)) %>% tibble::rownames_to_column("symbol") %>%
          .[,c("ens_id","symbol")] %>% as.data.table %>% setnames(c("symbol"),c("gene")),
        by=c("gene")
        )
rna_dt[,c("ens_id","gene"):=list(as.factor(ens_id),as.factor(gene))]
```

<!-- Merge data and sample metadata -->
```{r}
met_dt <- met_dt %>% merge(sample_metadata[,c("sample","lineage")], by="sample")
rna_dt <- rna_dt %>% merge(sample_metadata[,c("sample","lineage")], by="sample")
```

<!-- Merge DNA methylation and RNA expression data -->
```{r}
metrna <- merge(met_dt, rna_dt, by=c("sample","ens_id","lineage"))
rm(met_dt,rna_dt)
```

<!-- Overlap data with windows -->
```{r, echo=FALSE}
tmp <- seq(from=0-opts$up, to=0+opts$down-opts$window, by=opts$slide)
foo <- data.table(window_center=tmp+(opts$window/2), rel_start=tmp, rel_end=tmp+opts$window)

bar <- foverlaps(x=foo %>% setkey(rel_start,rel_end),
                 y=metrna[,c("rel_start","rel_end"):=dist ] %>% setkey(rel_start,rel_end)) %>%
  .[,c("i.rel_start","i.rel_end","rel_start","rel_end"):=NULL] %>%
  .[,.(rate=mean(rate), N=.N), by=c("sample","ens_id","gene","window_center","expr","lineage")]
rm(tmp,foo,metrna)
```

<!-- Compute local correlations -->
```{r}
cor <- bar %>% 
  .[,.(r=cor(x=rate, y=expr, method="pearson", use="pairwise.complete.obs")),
    by=c("sample","window_center","lineage")]
```

<!-- Plot -->
```{r}
p <- ggplot(cor,aes(x=window_center, y=r, fill=lineage, color=lineage)) +
  stat_summary(fun.data="mean_se", geom="ribbon", alpha=0.25) +
  stat_summary(geom="line") +
  geom_hline(yintercept=0, linetype="dashed", color="black", size=0.5) +
  geom_vline(xintercept=0, linetype="dashed", color="black", size=0.5) +
  scale_fill_manual(values=opts$colors) +
  scale_color_manual(values=opts$colors) +
  guides(fill=FALSE, color=FALSE) +
  ylab("Met/RNA Correlation") + xlab("Genomic distance from TSS (bp)") +
  scale_x_continuous(limits=c(-opts$up,opts$down)) +
  theme(
    axis.text.x = element_text(size=rel(1.1),colour="black"),
    axis.text.y = element_text(size=rel(1.1), colour="black"),
    axis.title.x = element_text(size=rel(1.2)),
    axis.title.y = element_text(size=rel(1.2)),
    legend.key = element_blank(),
    legend.position = "top",
    legend.direction = "horizontal",
    legend.justification = "center",
    legend.key.width = unit(2.0,"line"),
    legend.key.height = unit(2.0,"line"),
    legend.title = element_blank(),
    legend.text = element_text(size=rel(1.3)),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank()
  )
p

pdf(file=paste0(io$outdir,"/metrna_localcoupling_EB.pdf"), width=5, height=3.5)
print(p)
dev.off()
```

