---
title: ""
---

```{r load_modules, echo=FALSE, include=FALSE}
library(RColorBrewer)
library(ggpubr)
```

```{r}
theme_pub <- function() {
  theme(
    axis.title.y = element_text(colour="black", size=rel(1.3)),
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    axis.text.y = element_text(colour="black",size=rel(1.3)),
    axis.ticks = element_line(colour="black"),
    legend.position = "top",
    legend.title = element_blank(),
    legend.direction = "horizontal",
    legend.key.width = unit(1.2,"line"),
    legend.key.height = unit(1.0,"line"),
    legend.text = element_text(size=15)
  )
}
```

```{r}
################
## Define I/O ##
################

io <- list()
if (grepl("ricard",Sys.info()['nodename'])) {
  io$basedir <- "/Users/ricard/data/gastrulation"
} else if (grepl("ebi",Sys.info()['nodename'])) {
  io$basedir <- "/hps/nobackup2/research/stegle/users/ricard/scnmt_gastrulation"
} else {
  stop("Computer not recognised")
}  
io$metadata <- paste0(io$basedir,"/sample_metadata.txt")
io$indir <- paste0(io$basedir,"/met/results/hemimethylation")
io$outdir <- paste0(io$basedir,"/met/results/hemimethylation/pdf")

####################
## Define options ##
####################

opts <- list()

# Define genomic contexts (use NULL for no genomic context filtering)
opts$annos <- c(
  "all",
  "genebody",
  "prom_2000_2000_cgi",
  "prom_2000_2000_noncgi",
  # "prom_2000_2000"
  "H3K27ac_distal_E7.5_Mes_intersect12",
  "H3K27ac_distal_E7.5_Ect_intersect12",
  "H3K27ac_distal_E7.5_End_intersect12",
  # "H3K27ac_distal_E7.5_Mes_intersect12_500",
  # "H3K27ac_distal_E7.5_Ect_intersect12_500",
  # "H3K27ac_distal_E7.5_End_intersect12_500",
  # "exons",
  # "introns",
  # "CGI",
  # "LINE",
  "LTR" = "LTR"
)
# opts$annos <- NULL
```

Load metadata
```{r}
metadata <- fread(io$metadata) %>%
  .[,stage_lineage:=paste(stage,lineage10x_2,sep="_")] %>%
  .[!is.na(lineage10x_2) & lineage10x_2!="ExE_ectoderm"] 

# metadata <- metadata[stage%in%c("E6.5","E7.5")]
```

Load pre-computed results

```{r read_stats, echo=FALSE}
stats <- fread(paste0(io$indir,"/hemimethylation_short.txt.gz")) %>% 
  .[anno%in%opts$annos] %>%
  merge(metadata, by="sample")
```

<!-- Plots -->

```{r}
to.plot <- stats %>%
  .[,.(hemimethylation=mean(hemimethylation), number_cpgs=sum(number_cpgs)),by=c("stage","stage_lineage","anno")]
```

```{r}
for (i in unique(to.plot$stage)) {
  p <- ggplot(to.plot[stage==i], aes(x=anno, y=hemimethylation)) +
    geom_bar(stat="identity", color="black", position="dodge") +
    # facet_grid(~stage_lineage, scales="free_x", space = "free_x") +
    facet_wrap(~stage_lineage, scales="fixed") +
    geom_hline(aes(yintercept=hemimethylation), data=to.plot[stage==i & anno=="all"], linetype="dashed") +
    # coord_cartesian(ylim=c(0,100)) +
    ylab("Hemimethylation") +
    theme_bw() +
    theme(
      axis.text.x = element_text(colour="black",size=rel(0.8), angle=90, hjust=1, vjust=0.5),
      # axis.text.x = element_text(colour="black",size=rel(1.2)),
      strip.background = element_blank(),
      strip.text = element_text(color="black", size=rel(1.2))
    )
  
  # pdf(sprintf("%s/hemimeth_%s.pdf",io$outdir,i), width=9, height=5, useDingbats = F)
  print(p)
  # dev.off()
}
```

```{r}
to.plot <- stats[number_cpgs>=30]
```

```{r}
for (i in unique(to.plot$stage)) {
  p <- ggplot(to.plot[stage==i], aes(x=anno, y=hemimethylation)) +
    geom_boxplot(alpha=0.75, fill="#F8766D", outlier.shape = NA) +
    # facet_grid(~stage_lineage, scales="free_x", space = "free_x") +
    facet_wrap(~stage_lineage, scales="fixed") +
    # geom_jitter(alpha=0.75, color="#F8766D", size=1.5) +
    # coord_cartesian(ylim=c(0,100)) +
    ylab("Hemimethylation") +
    theme_bw() +
    theme(
      axis.text.x = element_text(colour="black",size=rel(0.8), angle=90, hjust=1, vjust=0.5),
      # axis.text.x = element_text(colour="black",size=rel(1.2)),
      strip.background = element_blank(),
      strip.text = element_text(color="black", size=rel(1.2))
    )
  
  # pdf(sprintf("%s/hemimeth_%s.pdf",io$outdir,i), width=9, height=5, useDingbats = F)
  print(p)
  # dev.off()
}
```

By cell cycle phase
```{r}
cell.cycle <- fread("/Users/ricard/data/gastrulation/rna/results/cell_cycle/cell_cycle_scran.txt.gz") %>%
  .[phase2%in%c("G1S","G2M")] %>%
  .[,c("id_rna","phase","phase2")]

table(cell.cycle$phase)
table(cell.cycle$phase2)

to.plot <- stats %>% 
  merge(cell.cycle,by="id_rna") %>%
  .[,.(hemimethylation=mean(hemimethylation), number_cpgs=sum(number_cpgs)),by=c("stage","stage_lineage","anno","phase")]
```

```{r}
for (i in unique(to.plot$stage)) {
  p <- ggplot(to.plot[stage==i], aes(x=anno, y=hemimethylation, fill=phase)) +
    geom_bar(stat="identity", color="black", position="dodge") +
    facet_wrap(~stage_lineage, scales="fixed") +
    ylab("Hemimethylation") +
    theme_bw() +
    theme(
      axis.text.x = element_text(colour="black",size=rel(0.8), angle=90, hjust=1, vjust=0.5),
      # axis.text.x = element_text(colour="black",size=rel(1.2)),
      strip.background = element_blank(),
      strip.text = element_text(color="black", size=rel(1.2))
    )
  
  # pdf(sprintf("%s/hemimeth_%s.pdf",io$outdir,i), width=9, height=5, useDingbats = F)
  print(p)
  # dev.off()
}
```

