---
title: "Gastrulation: Coupling between RNA expression and DNA methylation"
output: 
  BiocStyle::html_document: 
    fig_width: 12
    fig_height: 8
---

```{r echo=FALSE, include=FALSE}
suppressPackageStartupMessages(library(scater))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(weights))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(data.table))
```

<!-- Define functions -->
```{r echo=FALSE, include=FALSE}
source("/Users/ricard/NMT-seq_ESC/correlations/utils.R")
```

<!-- Define I/O and options -->
```{r echo=TRUE, include=FALSE}
# Data
io   <- list()
io$base_dir   <- "/Users/ricard/data/gastrulation"
io$in.sample_metadata <- paste0(io$base_dir,"/sample_metadata_scNMT.txt")
io$annos_dir  <- paste0(io$base_dir, "/features/filt")
io$rna_file   <- paste0(io$base_dir, "/rna/parsed/sceset_scNMT.rds")
io$met_dir   <- paste0(io$base_dir, "/met/parsed")
io$outdir <- "/Users/ricard/gastrulation/metrna/coupling/out"
io$gene_metadata <- "/Users/ricard/data/ensembl/mouse/v87/BioMart/mRNA/Mmusculus_genes_BioMart.87.txt"

# Filtering parameters
opts <- list()
opts$min.CpGs <- 3       # Minimum number of CpGs per feature
opts$gene_window <- 1e4  # window length for the overlap between genes and features

# Multiple testing options
opts$threshold_fdr  <- 0.01   # pvalue threshold for significance in FDR

# Correlation type options
opts$method <- "pearson"      # correlation type
opts$weight <- FALSE          # weighted correlation? 

# Permutation test options
opts$permutation <- TRUE   # do permutation test?
opts$n_perms <- 10         # Number of permutations

# Define genomic contexts
opts$annos <- c(
  "genebody",
  # "prom_2000_2000_cgi",
  "prom_2000_2000_noncgi",
  # "prom_2000_2000",
  "CGI",
  # "H3K27ac_distal_E7.5_union_500",
  # "H3K27ac_promoter_E7.5_union_500",
  # "H3K4me3_E7.5_union",
  "LINE"="LINE",
  "LTR"="LTR"
)

# Define stage and lineages
opts$stage_lineage <- c(
  "E4.5_EPI",#"E4.5_PE",
  "E5.5_EPI",#"E5.5_PE",
  "E6.5_EPI","E6.5_PS",#"E6.5_VE",
  "E7.5_Ectoderm","E7.5_Mesoderm","E7.5_Endoderm"
)

# Define which cells to use
foo <- fread(io$in.sample_metadata) %>% 
  .[,stage_lineage:=paste(stage,lineage,sep="_")] %>%
  .[pass_metQC==T & pass_rnaQC==T & stage_lineage%in%opts$stage_lineage] 
  # .[pass_metQC==T & pass_rnaQC==T & stage_lineage%in%opts$stage_lineage & KO_3b=="not"] 
opts$rna.cells <- foo$id_rna
opts$met.cells <- foo$id_met
```

<!-- Load sample metadata -->
```{r}
sample_metadata <- fread(io$in.sample_metadata, stringsAsFactors=T) %>% 
  .[,c("id_rna","id_met","sample","stage","lineage")] %>%
  .[,stage_lineage:=paste(stage,lineage,sep="_")] %>%
  .[sample%in%foo$sample]
```

<!-- Load RNA data -->
```{r load_rna, echo=FALSE, include=FALSE}
# Load expression as SCESet object
sce <- readRDS(file = io$rna_file)

# Subset cells
sce <- sce[,colnames(sce) %in% opts$rna.cells]

# Create data.table
rna_dt <- exprs(sce) %>% t %>% as.data.table(keep.rownames="id_rna") %>% 
  melt(id.vars="id_rna", value.name="expr", variable.name="id") %>%
  merge(sample_metadata[,c("sample","id_rna")]) %>%
  merge(rowData(sce) %>% as.data.frame(row.names=rownames(sce)) %>% tibble::rownames_to_column("ens_id") %>%
          .[,c("symbol","ens_id")] %>% as.data.table %>% setnames(c("ens_id","symbol"),c("id","gene")),
        by=c("id")
        )
rna_dt[,c("id","gene"):=list(as.factor(id),as.factor(gene))]

# Extract gene coordinates metadata to do the overlap
# rna_metadata <- rowRanges(sce) %>% as.data.frame %>% tibble::rownames_to_column("ens_id") %>% as.data.table %>% 
#   .[,c("seqnames","start","end","symbol","ens_id")] %>% setnames(c("seqnames","ens_id","symbol"), c("chr","id","gene")) %>%
#   .[, chr := sub("chr", "", chr)]
# rna_metadata[,c("chr","id","gene"):=list(as.factor(chr),as.factor(id),as.factor(gene))]
```

<!-- Load methylation data -->
```{r load_met, echo=FALSE, include=FALSE}
met_dt <- lapply(opts$annos, function(n) {
  data <- fread(sprintf("zcat < %s/%s.tsv.gz",io$met_dir,n), showProgress=F, stringsAsFactors=T, sep="\t", quote="") %>%
  .[V1%in%opts$met.cells]
}) %>% rbindlist
colnames(met_dt) <- c("sample","id","anno","Nmet","N","rate")
```

<!-- Load genomic contexts -->
```{r load_genomiccontexts}
feature_metadata <- lapply(opts$annos, function(anno) 
  fread(sprintf("%s/%s.bed",io$annos_dir,anno), stringsAsFactors=T)[,c(1,2,3,4,5,6)]) %>%
  rbindlist %>% setnames(c("chr","start","end","strand","id","anno"))

# Load gene metadata 
gene_metadata <- fread(io$gene_metadata,stringsAsFactors=T) %>% 
  setnames(c("ens_id","symbol"),c("id","gene")) %>% 
  .[,chr:=as.factor(substr(chr,4,4))]
```

<!-- Parse gene and feature metadata -->
```{r}
feature_metadata_filt <- feature_metadata %>% split(.$anno) %>% 
  map2(.,names(.), function(x,y) x[id %in% met_dt[anno==y,id]] ) %>%
  rbindlist

gene_metadata_filt <- gene_metadata %>% .[,c("chr","start","end","gene")] %>% 
  .[,c("start", "end") := list(start-opts$gene_window, end+opts$gene_window)] %>% 
  setkey(chr,start,end)
```

<!-- Associate the non-genic contexts with overlapping genes -->
```{r associate, echo=FALSE}
met_list <- list()
for (ann in unique(met_dt$anno)){
  
  # Subset corresponding anno
  met_tmp <- met_dt[anno == ann, ]
  
  # Non gene-associated feature
  if (all(grepl("ENSMUSG", unique(met_tmp$id)) == FALSE)) {
    
    # Extract coordiantes for methylation sites and for genes
    feature_metadata_tmp <- feature_metadata_filt[anno==ann, c("chr","start","end","id")] %>% 
      .[,c("start","end") := list(start - opts$gene_window, end + opts$gene_window)] %>% setkey(chr,start,end)
    
    # Do the overlap
    ov <- foverlaps(gene_metadata_filt, feature_metadata_tmp, nomatch=0) %>% .[,c("gene", "id")]# %>%
      # # If a feature overlap with more than one gene, concatenate all of them
      # .[,.(gene=paste(gene, collapse="_")),by="id"] %>%
      # .[,c("gene", "id")]
    
    # Merge with methylation data
    met_list[[ann]] <- merge(met_tmp, ov, by="id", allow.cartesian=T) 
  }
  # Gene-associated feature
  else if (all(grepl("ENSMUSG", unique(met_tmp$id)) == TRUE)) {
    met_list[[ann]] <- merge(met_tmp, gene_metadata[,c("id","gene")], by="id")
  }
}
met_dt <- rbindlist(met_list)
rm(met_list,met_tmp,ov,feature_metadata_tmp)
```

<!-- Merge DNA methylation and RNA expression data -->
```{r}
metrna_dt <- merge(
  met_dt, 
  rna_dt[,c("sample","gene","expr")], 
  by=c("sample","gene")
)
```

<!-- Filter data -->
```{r filter_all, echo=FALSE}
# Filter by mininum number of CpGs per feature
metrna_dt <- metrna_dt[N >= opts$min.CpGs]
```

<!-- Compute coupling -->

Option 1: Binarise RNA expression and Methylation rate and do a Fisher test of equal proportion
```{r}
# binarised_metrna_dt <- metrna_dt %>% copy %>% 
#   .[,expr_binary:=ifelse(expr>0,"Expr","Non-Expr")] %>%
#   .[,rate_binary:=ifelse(rate>50,"Met","Non-Met")] %>%
#   .[,c("expr_binary","rate_binary"):=list(as.factor(expr_binary),as.factor(rate_binary))]
#   
# cor <- binarised_metrna_dt %>%
#   .[,.(n=.N, V1=unlist(fisher.test(as.factor(expr_binary),as.factor(rate_binary))[c("estimate","p.value")])), by=c("sample","anno")] %>%
#   .[,para := c("odds","p")] %>% dcast(sample+anno~para,value.var="V1") %>% 
#   .[, c("padj_fdr","padj_bonf") := list(p.adjust(p, method="fdr"),p.adjust(p, method="bonferroni")), by=anno] %>%
#   .[, c("log_padj_fdr", "log_padj_bonf") := list(-log10(padj_fdr), -log10(padj_bonf))] %>%
#   .[, sig := padj_fdr <= opts$threshold_fdr] %>% setorder(padj_fdr)
# 
# for (n in unique(cor$anno)) {
#   tmp <- merge(cor[anno==n], sample_metadata[,c("sample","stage","KO_3b")], by="sample")
#   p <- ggplot(tmp, aes(x=stage, y=odds, fill=interaction(stage,KO_3b))) +
#     geom_boxplot(alpha=0.8, outlier.shape=NA) +
#     geom_hline(yintercept=1, colour="black", linetype="dashed") +
#     labs(title=n, x="", y="Odds ratio") +
#     boxplot_theme() + scale_fill_brewer(palette = "Accent") +
#     theme(
#       axis.text.x = element_blank()
#     )
#   print(p)
#   
# }


# Odds:
# - Expressed given methylation: Expr-Met / NonExpr-Met
# - Expressed given unmethylated: Expr-UnMet / NonExpr-UnMet
# (DE/HE) / (DN/HN)
# test <- binarised_metrna_dt[sample=="E7.5_Plate4_H9" & anno=="prom_2000_2000_noncgi"]
# asd <- table(test$expr_binary, test$rate_binary)
# asd
# 
# odds1 <- asd["Expr","Met"] / asd["Non-Expr","Met"]
# odds2 <- asd["Expr","Non-Met"] / asd["Non-Expr","Non-Met"]
# or <- odds2/odds1
# or
```

Option 2: Continuous Pearson correlation coefficient
```{r cor_rate, echo=FALSE, include=FALSE}
# Weighted correlation
if (opts$weight == TRUE){
  if (opts$method != "pearson") { print("Weighted correlation only supported for pearson"); stop() }
  cor <- metrna_dt[, wtd.cor(rate, expr, N)[, c("correlation", "t.value", "p.value")], by = c("sample", "anno")]
  
# Unweighted correlation
} else{
  cor <- metrna_dt[, .(V1 = unlist(cor.test(rate, expr, alternative = "two.sided", method = opts$method)[c("estimate", "statistic", "p.value")])), by = c("sample", "anno")]
}

# Compute adjusted p-values (both FDR and Bonferroni)
cor <- cor %>% .[,para := c("r", "t", "p")] %>% dcast(sample + anno ~ para, value.var = "V1") %>% 
      .[, c("padj_fdr", "padj_bonf") := list(p.adjust(p, method = "fdr"), p.adjust(p, method = "bonferroni")), by = anno] %>%
      .[, c("log_padj_fdr", "log_padj_bonf") := list(-log10(padj_fdr), -log10(padj_bonf))] %>%
      .[, sig := padj_fdr <= opts$threshold_fdr] %>% setorder(padj_fdr)
```

<!-- Save results -->
```{r}
fwrite(cor, paste0(io$outdir,"/metrna_coupling.txt"), sep="\t")
cor <- fread(paste0(io$outdir,"/metrna_coupling.txt"))
```

<!-- Run permutation test -->
```{r}
if (opts$permutation) {
  pp_vals <- vector(mode = "numeric", length = length(cor$p))
  for (k in 1:opts$n_perms){
    print(k)
    metrna_dt_perm <- copy(metrna_dt)
    
    # Permute gene expression levels and methylation rates
    metrna_dt_perm[, c("rate","expr") := list(sample(rate),sample(expr)), by = c("sample","anno")]
    
    # Compute coupling
    cor_perm <- metrna_dt_perm[, .(p = cor.test(rate, expr, method=opts$method)[["p.value"]]), by = c("sample","anno")]
    
    # For each annotation sort by p-value and store them
    cor_perm <- cor_perm %>% split(.$anno) %>% map(~ .[,.(anno = anno, p = sort(p))]) %>% rbindlist
    pp_vals <- pp_vals + cor_perm %>% .[, p]
  }
  # Compute the average p-values across all permutations
  pp_vals <- pp_vals / opts$n_perms
  cor_perm <- cor_perm[, p := pp_vals]
}
```

<!-- QQplots of true vs permuted signal -->
```{r, echo=FALSE}
if (opts$permutation) {
  pp <- qq <- list()
  for (n in unique(cor$anno)) {
    
    # Generate permutation plot
    qq[[n]] <- gg_qqplot(cor[anno == n, ], cor_perm[anno == n]$p, title = "")
    
    # Combine and save
    # pdf(file=paste0(io$outdir,"/permutation_", n, ".pdf"), height=4, width=6, useDingbats = FALSE)
    print(qq[[n]])
    # dev.off()
    
  }
}
```

<!-- Boxplots -->

Annotation in the X axis and grouped by stage
```{r}
# Divergent
opts$colors <- c(E4.5="#F8766D", E5.5="#7CAE00", E6.5="#00BFC4", E7.5="#C77CFF")

# Methylation progression
# opts$colors <- c(E4.5="#FFC400", E5.5="#FF5627", E6.5="#C90035", E7.5="#5B1647")

subset_annos <- c(
  "prom_2000_2000_noncgi"="Promoters",
  # "prom_2000_2000"="foo Promoters",
  "genebody"="Gene\nbodies",
  # "H3K27ac_distal_E7.5_common_500"="Distal\nH3K27ac",
  # "H3K27ac_promoter_E7.5_common_500"="Proximal\nH3K27ac",
  # "H3K27ac_distal_E7.5_union_500"="E7.5 Distal\nH3K27ac",
  # "H3K27ac_promoter_E7.5_union_500"="E7.5 Proximal\nH3K27ac",
  # "H3K4me3_E7.5_common"="H3K4me3",
  # "H3K4me3_E7.5_union"="E7.5\nH3K4me3",
  # "ESC_p300"="p300 ChIP-seq (ESCs)",
  # "ESC_CTCF"="CTCF ChIP-seq (ESCs)",
  "CGI"="CpG\nislands",
  "LINE"="LINE",
  "LTR"="LTR"
)
  

tmp <- merge(cor, sample_metadata[,c("sample","stage","stage_lineage")], by="sample") %>%
  .[anno%in%names(subset_annos)] %>%
  .[,anno:=stringr::str_replace_all(anno, subset_annos)] %>%
  .[,anno:=factor(anno,levels=subset_annos)]


# p <- ggplot(tmp, aes(x=anno, y=r, fill=stage)) +
p <- ggplot(tmp, aes(x=anno, y=r, fill=stage)) +
  geom_boxplot(alpha=0.8, outlier.shape=NA) +
  geom_hline(yintercept=0, colour="black", linetype="dashed") +
  scale_fill_manual(values=opts$colors) +
  # coord_cartesian(ylim=c(-0.25,0.1)) +
  labs(title="", x="", y="Met/RNA correlation") +
  boxplot_theme() +
  # scale_fill_brewer(palette = "Accent") +
  theme(
    legend.position=c(0.3,1.05),
    axis.text.x = element_text(color="black", size=11)
  )

pdf(paste0(io$outdir,"/metrna_coupling_boxplots2.pdf"), width=9, height=5, useDingbats = F)
print(p)
dev.off()
```

<!-- Compute statistics on the stage transitions -->
```{r}
stage_transitions <- c(
  "E4.5"="E5.5",
  "E5.5"="E6.5",
  "E6.5"="E7.5"
  )

subset_annos <- c("genebody"="Gene\nbodies", "CGI"="CpG\nislands", "LINE"="LINE", "LTR"="LTR")

tmp <- merge(cor, sample_metadata[,c("sample","stage","stage_lineage")], by="sample") %>%
  .[anno%in%names(subset_annos)] %>%
  .[,anno:=stringr::str_replace_all(anno, subset_annos)] %>%
  .[,anno:=factor(anno,levels=subset_annos)]

for (stage1 in names(stage_transitions)) {
  for (i in subset_annos) {
    stage2 <- stage_transitions[stage1]
    foo1 <- tmp[stage==stage1 & anno==i]
    foo2 <- tmp[stage==stage2 & anno==i]
    p.value <- t.test(foo1$r,foo2$r)$p.value
    print(sprintf("%s to %s, %s: %.10f\n",stage1,stage2,i,p.value))
  }
  cat("\n")
}
```


