---
title: "Gastrulation: genome-wide DNA methylation distributions"
output:
  BiocStyle::html_document: 
    toc: false
    fig_width: 10
    fig_height: 8
---

```{r load_modules, echo=FALSE, include=FALSE}
library(data.table)
library(purrr)
library(ggplot2)
```

```{r define_opts, echo=FALSE}

## Define I/O ##
io <- list()
io$sample.metadata <- "/Users/ricard/data/gastrulation/sample_metadata.txt"
io$data <- "/Users/ricard/data/gastrulation/met/raw"

## Define options ##
opts <- list()

# Define which cells to use (use NULL for all cells)
opts$stage_lineage10x <- c(
  "E4.5_Epiblast"
  # "E4.5_Primitive_endoderm",
  # "E5.5_Epiblast",
  # "E5.5_Visceral_endoderm",
  # "E6.5_Epiblast",
  # "E6.5_Primitive_Streak",
  # # "E6.5_Mesoderm",
  # "E6.5_Visceral_endoderm",
  # "E7.5_Primitive_Streak",
  # "E7.5_Endoderm",
  # "E7.5_Mesoderm",
  # "E7.5_Epiblast",
  # "E7.5_Ectoderm"
)
# opts$stage_lineage10x <- NULL

# Define which cells to use
tmp <- fread(io$sample.metadata) %>% .[!is.na(id_met)] 
if (is.null(opts$stage_lineage10x)) {
  opts$cells <- tmp[,id_met]
} else {
  opts$cells <- tmp %>%
    .[,stage_lineage:=paste(stage,lineage10x_2,sep="_")] %>%
    .[pass_metQC==T & stage_lineage%in%opts$stage_lineage,id_met]
}

# Define which chromosomes to use
opts$chr <- paste0("chr",as.character(1:19))
```

<!-- load metadata -->
```{r}
# Load sample metadata
sample_metadata <- fread(io$sample.metadata,stringsAsFactors=F) %>% 
  .[,c("id_met","stage","lineage10x_2","plate")] %>%
  .[id_met%in%opts$cells] %>%
  .[,stage_lineage:=as.factor(paste(stage,lineage10x_2,sep="_"))]

# Load genomic coordinates
# mm10 <- fread("/Users/ricard/data/mm10_sequence/mm10.genome") %>% 
#   setnames(c("chr","length")) %>% 
#   .[chr %in% opts$chr]
```

<!-- Load DNA methylation rate per cell  -->
```{r}
# data_list <- list()
# for (i in 1:length(opts$cells)) {
#   data_list[[i]] <- fread(sprintf("zcat < %s/%s.tsv.gz",io$in.data, opts$cells[i])) %>% 
#     .[,c("chr","pos","rate")] %>%
#     .[,id_met:=as.factor(opts$cells[i])] %>% 
#     .[,chr:=as.factor(paste0("chr",chr))]
# }
# data <- rbindlist(data_list)

plan(multiprocess)

data <- future_lapply(1:length(opts$cells), function(i) {
    fread(sprintf("zcat < %s/%s.tsv.gz",io$in.data, opts$cells[i])) %>% 
    .[,c("chr","pos","rate")] %>%
    .[,id_met:=as.factor(opts$cells[i])] %>% 
    .[,chr:=as.factor(paste0("chr",chr))]
}) %>% rbindlist
```

<!-- Parse data -->
```{r}
# Merge methylation data and sample metadata
data <- data %>% merge(sample_metadata[,c("sample","stage_lineage")], by="sample")

# Convert beta value to M value
data[,m:=log2(((rate/100)+0.01)/(1-(rate/100)+0.01))]
```
